psdisk/psdisk.c

```diff
@@ -12,156 +12,570 @@
+static void psdisk_showPartsTable(void)
+{
+	int sz;
+	struct plist_node_t *node;
+
+	fseek(psdisk_common.file, 0L, SEEK_END);
+	sz = ftell(psdisk_common.file);
+
+	printf("\n");
+	printf(BOLDWHITE "File system %s: %d bytes\n"RESET, psdisk_common.fileName, sz);
+	printf("Memory size %u bytes\n", psdisk_common.mem.memSize);
+	printf("Units: sector %u bytes \n", psdisk_common.mem.sectorSize);
+	printf("\n");
+	printf(BOLDWHITE PARTITION_HEADER_ALIGMENT RESET, "Name", "Start", "End", "Sectors", "Size", "Type");
+	LIST_FOREACH(node, &psdisk_common.list, ptrs) {
+		printf(PARTITION_DATA_ALIGMENT, node->pHeader.name,
+										node->pHeader.offset,
+										node->pHeader.offset + node->pHeader.size,
+										node->pHeader.size / psdisk_common.mem.sectorSize,
+										node->pHeader.size,
+										psdisk_getTypeName(node->pHeader.type));
+	}
+	printf("\n");
+}
+
+
+/* Callbacks to ptable library */
+
+static ssize_t psdisk_readImg(unsigned int addr, void *buff, size_t bufflen)
+{
+	int offs;
+
+	if (PTABLE_WITH_OFFSET(psdisk_common.opts))
+		offs = addr;
+	else
+		offs = addr - (psdisk_common.mem.memSize - psdisk_common.mem.sectorSize);
+
+	if (fseek(psdisk_common.file, offs, SEEK_SET) != 0) {
+		printf("Cannot change file position, err: %s.\n", strerror(errno));
+		return -1;
+	}
+
+	return fread(buff, sizeof(uint8_t), bufflen, psdisk_common.file);
+}
+
+
+static ssize_t psdisk_writeImg(unsigned int addr, const void *buff, size_t bufflen)
+{
+	ssize_t res;
+	char *wBuff;
+
+	if (fseek(psdisk_common.file, 0, SEEK_SET) != 0) {
+		printf("Cannot change file position, err: %s.\n", strerror(errno));
+		return -1;
+	}
+
+	if ((wBuff = (char *)malloc(psdisk_common.mem.sectorSize)) == NULL) {
+		printf("Cannot allocate memory, err: %s.\n", strerror(errno));
+		return -1;
+	}
+
+	memset(wBuff, 0xff, psdisk_common.mem.sectorSize);
+
+	if (PTABLE_WITH_OFFSET(psdisk_common.opts)) {
+		while (addr > 0) {
+			fwrite(wBuff, sizeof(uint8_t), psdisk_common.mem.sectorSize, psdisk_common.file);
+			addr -= psdisk_common.mem.sectorSize;
+		}
+	}
+
+	memcpy(wBuff, buff, bufflen);
+	res = fwrite(wBuff, sizeof(uint8_t), psdisk_common.mem.sectorSize, psdisk_common.file);
+
+	free(wBuff);
+
+	return res;
+}
+
+
+/* Partition table modification functions */
+
+static int psdisk_readPartsTable(void)
 {
-	/* TO DO: add option to move partition table at the end of an image ?? */
-	return fwrite(buff, sizeof(uint8_t), bufflen, psdisk_common.file);
+	int i;
+	uint32_t pCnt;
+	struct plist_node_t *pNode;
+	ptable_partition_t *pHeaders;
+
+	if ((pHeaders = ptable_readPartitions(&pCnt, &psdisk_common.mem)) == NULL) {
+		printf("The file contains incorrect partition table.\n");
+		return -1;
+	}
+
+	for (i = 0; i < pCnt; ++i) {
+		if ((pNode = (struct plist_node_t *)calloc(1, sizeof(struct plist_node_t))) == NULL) {
+			printf("Cannot allocate memory, err: %s.\n", strerror(errno));
+			free(pHeaders);
+			return -1;
```

Comment:
I think we should destroy and free list and it's members here.

Reply:
Destroying list and cleaning up is located in main function or in `psdisk_handleImg`. I wanted to avoid mess in code invoking `psdisk_destroy` in many functions. In this case there is no need to destroy and free list's members, it is done after return from this function.